In [1]:
%pip install -q kagglehub
import kagglehub
import pandas as pd
import os

path = kagglehub.dataset_download(
    "olistbr/brazilian-ecommerce"
)

print(path)


[notice] A new release of pip is available: 25.2 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.
C:\Users\Usuario\.cache\kagglehub\datasets\olistbr\brazilian-ecommerce\versions\2


In [2]:
files = os.listdir(path)

for file in files:
    print(file)

olist_customers_dataset.csv
olist_geolocation_dataset.csv
olist_orders_dataset.csv
olist_order_items_dataset.csv
olist_order_payments_dataset.csv
olist_order_reviews_dataset.csv
olist_products_dataset.csv
olist_sellers_dataset.csv
product_category_name_translation.csv


In [3]:
#asi se cargan las tablas en un diccionario, donde la clave es el nombre de la tabla y el valor es el DataFrame

tables = {}
for file in files:
    if file.endswith(".csv"):
        table_name = file.replace(".csv", "")
        table_path = os.path.join(path, file)
        tables[table_name] = pd.read_csv(table_path)
        print(f"Loaded table: {table_name}")

tables["olist_customers_dataset"].head()

Loaded table: olist_customers_dataset
Loaded table: olist_geolocation_dataset
Loaded table: olist_orders_dataset
Loaded table: olist_order_items_dataset
Loaded table: olist_order_payments_dataset
Loaded table: olist_order_reviews_dataset
Loaded table: olist_products_dataset
Loaded table: olist_sellers_dataset
Loaded table: product_category_name_translation


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [4]:
#ahora vamos a ver el tamanio de las tablas
for name, df in tables.items():
    print(f"Table: {name}, Rows: {df.shape[0]}, Columns: {df.shape[1]}")

Table: olist_customers_dataset, Rows: 99441, Columns: 5
Table: olist_geolocation_dataset, Rows: 1000163, Columns: 5
Table: olist_orders_dataset, Rows: 99441, Columns: 8
Table: olist_order_items_dataset, Rows: 112650, Columns: 7
Table: olist_order_payments_dataset, Rows: 103886, Columns: 5
Table: olist_order_reviews_dataset, Rows: 99224, Columns: 7
Table: olist_products_dataset, Rows: 32951, Columns: 9
Table: olist_sellers_dataset, Rows: 3095, Columns: 4
Table: product_category_name_translation, Rows: 71, Columns: 2


In [5]:
#primer resumen basico automatico, seria un data quality overview, que nos da una idea de la calidad de los datos y de la estructura de las tablas
summary = {}

for name, df in tables.items():
    summary[name] = {
        "rows": df.shape[0],
        "columns": df.shape[1],
        "missing_values": df.isnull().sum().sum(),
        "duplicate_rows": df.duplicated().sum(),
        "data_types": df.dtypes.value_counts().to_dict()
    }
summary_df = pd.DataFrame.from_dict(summary, orient='index')
summary_df

,rows,columns,missing_values,duplicate_rows,data_types
olist_customers_dataset,99441,5,0,0,"{str: 4, int64: 1}"
olist_geolocation_dataset,1000163,5,0,261831,"{float64: 2, str: 2, int64: 1}"
olist_orders_dataset,99441,8,4908,0,{str: 8}
olist_order_items_dataset,112650,7,0,0,"{str: 4, float64: 2, int64: 1}"
olist_order_payments_dataset,103886,5,0,0,"{str: 2, int64: 2, float64: 1}"
olist_order_reviews_dataset,99224,7,145903,0,"{str: 6, int64: 1}"
olist_products_dataset,32951,9,2448,0,"{float64: 7, str: 2}"
olist_sellers_dataset,3095,4,0,0,"{str: 3, int64: 1}"
product_category_name_translation,71,2,0,0,{str: 2}


In [6]:
orders = tables["olist_orders_dataset"].copy()
orders.info()
orders.head()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 21.9 MB


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [7]:
nulls = pd.DataFrame({
    "columna": orders.columns,
    "cantidad_nulos": orders.isnull().sum(),
    "porcentaje_nulos": orders.isnull().mean() * 100
})

nulls

# condicion = orders["order_approved_at"].isnull() & orders["order_delivered_carrier_date"].notnull()
# orders[condicion][["order_id", "order_approved_at", "order_delivered_carrier_date", "order_purchase_timestamp"]]

# condicion = orders["order_approved_at"].isnull()
# approved_nulls = orders[condicion]
# approved_nulls.head()

# print(f"Cantidad de filas con order_approved_at nulo: {approved_nulls.shape[0]}")
# approved_nulls["order_status"].value_counts()

# condicion2 = approved_nulls["order_status"] == "delivered"
# delivered_nulls = approved_nulls[condicion2]
# print(f"Cantidad de filas con order_approved_at nulo y order_status 'delivered': {delivered_nulls.shape[0]}")
# delivered_nulls[["order_id", "order_approved_at", "order_delivered_carrier_date", "order_purchase_timestamp"]]


,columna,cantidad_nulos,porcentaje_nulos
order_id,order_id,0,0.000000
customer_id,customer_id,0,0.000000
order_status,order_status,0,0.000000
order_purchase_timestamp,order_purchase_timestamp,0,0.000000
order_approved_at,order_approved_at,160,0.160899
order_delivered_carrier_date,order_delivered_carrier_date,1783,1.793023
order_delivered_customer_date,order_delivered_customer_date,2965,2.981668
order_estimated_delivery_date,order_estimated_delivery_date,0,0.000000


In [8]:
condicion_created = (
    approved_nulls["order_status"] == "created"
)
created_nulls = approved_nulls[condicion_created]
print(f"Cantidad de filas con order_approved_at nulo y order_status 'created': {created_nulls.shape[0]}")
created_nulls[
    [
        "order_id",
        "order_status",
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
]

NameError: name 'approved_nulls' is not defined

In [ ]:
delivered_nulls[
    ["order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"]
]

,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
5323,2017-02-18 14:40:00,NaN,2017-02-23 12:04:47,2017-03-01 13:25:33,2017-03-17 00:00:00
16567,2017-02-18 12:45:31,NaN,2017-02-23 09:01:52,2017-03-02 10:05:06,2017-03-21 00:00:00
19031,2017-02-18 13:29:47,NaN,2017-02-22 16:25:25,2017-03-01 08:07:38,2017-03-17 00:00:00
22663,2017-02-18 16:48:35,NaN,2017-02-22 11:23:10,2017-03-09 07:28:47,2017-03-31 00:00:00
23156,2017-02-17 13:05:55,NaN,2017-02-22 11:23:11,2017-03-02 11:09:19,2017-03-20 00:00:00
26800,2017-01-19 12:48:08,NaN,2017-01-25 14:56:50,2017-01-30 18:16:01,2017-03-01 00:00:00
38290,2017-02-19 01:28:47,NaN,2017-02-23 03:11:48,2017-03-02 03:41:58,2017-03-27 00:00:00
39334,2017-02-18 11:04:19,NaN,2017-02-23 07:23:36,2017-03-02 16:15:23,2017-03-22 00:00:00
48401,2017-01-19 22:26:59,NaN,2017-01-27 11:08:05,2017-02-06 14:22:19,2017-03-16 00:00:00
61743,2017-02-17 17:21:55,NaN,2017-02-22 11:42:51,2017-03-03 12:16:03,2017-03-20 00:00:00


Auditoria de missing values:
De los 160 pedidos sin fecha de aprobación, 141 están cancelados y 5 permanecen en estado created, por lo que la ausencia puede ser coherente con su flujo. Sin embargo, existen 14 pedidos con estado delivered que sí poseen fechas de despacho y entrega, pero carecen de order_approved_at, lo que indica una inconsistencia o pérdida de información en ese atributo.

In [ ]:
carrier_nulls = orders[orders["order_delivered_carrier_date"].isna()]
carrier_nulls["order_status"].value_counts()

carrier_nulls[carrier_nulls["order_status"] == "delivered"]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
73222,2aa91108853cecb43c84a5dc5b277475,afeb16c7f46396c0ed54acb45ccaaa40,delivered,2017-09-29 08:52:58,2017-09-29 09:07:16,NaN,2017-11-20 19:44:47,2017-11-14 00:00:00
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaN,NaN,2017-06-23 00:00:00


Los valores nulos de order_delivered_carrier_date son, en su mayoría, consistentes con la lógica del proceso de Olist, ya que corresponden a pedidos que fueron cancelados, no estuvieron disponibles o todavía no alcanzaron la etapa de envío. Sin embargo, se detectaron dos pedidos con estado delivered y sin fecha de entrega al transportista. Uno posee fecha de entrega al cliente, lo que constituye una inconsistencia evidente; el otro tampoco posee fecha de entrega al cliente, por lo que no es posible determinar si el error se encuentra en el estado del pedido o en las fechas registradas. Debido a la falta de evidencia para reconstruir estas fechas, se mantienen los valores nulos y se consideran anomalías de calidad de datos.


In [ ]:
delivered_customer_nulls = orders[orders["order_delivered_customer_date"].isna()]
delivered_customer_nulls["order_status"].value_counts()
delivered_customer_nulls[delivered_customer_nulls["order_status"] == "delivered"]

delivered_customer_anomalies = orders[
    (orders["order_status"] == "delivered") &
    (orders["order_delivered_customer_date"].isna())
]

delivered_customer_anomalies


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
3002,2d1e2d5bf4dc7227b3bfebb81328c15f,ec05a6d8558c6455f0cbbd8a420ad34f,delivered,2017-11-28 17:44:07,2017-11-28 17:56:40,2017-11-30 18:12:23,NaN,2017-12-18 00:00:00
20618,f5dd62b788049ad9fc0526e3ad11a097,5e89028e024b381dc84a13a3570decb4,delivered,2018-06-20 06:58:43,2018-06-20 07:19:05,2018-06-25 08:05:00,NaN,2018-07-16 00:00:00
43834,2ebdfc4f15f23b91474edf87475f108e,29f0540231702fda0cfdee0a310f11aa,delivered,2018-07-01 17:05:11,2018-07-01 17:15:12,2018-07-03 13:57:00,NaN,2018-07-30 00:00:00
79263,e69f75a717d64fc5ecdfae42b2e8e086,cfda40ca8dd0a5d486a9635b611b398a,delivered,2018-07-01 22:05:55,2018-07-01 22:15:14,2018-07-03 13:57:00,NaN,2018-07-30 00:00:00
82868,0d3268bad9b086af767785e3f0fc0133,4f1d63d35fb7c8999853b2699f5c7649,delivered,2018-07-01 21:14:02,2018-07-01 21:29:54,2018-07-03 09:28:00,NaN,2018-07-24 00:00:00
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaN,NaN,2017-06-23 00:00:00
97647,ab7c89dc1bf4a1ead9d6ec1ec8968a84,dd1b84a7286eb4524d52af4256c0ba24,delivered,2018-06-08 12:09:39,2018-06-08 12:36:39,2018-06-12 14:10:00,NaN,2018-06-26 00:00:00
98038,20edc82cf5400ce95e1afacc25798b31,28c37425f1127d887d7337f284080a0f,delivered,2018-06-27 16:09:12,2018-06-27 16:29:30,2018-07-03 19:26:00,NaN,2018-07-19 00:00:00


Se analizaron los valores nulos de order_delivered_customer_date. La mayoría son coherentes con la lógica del proceso de negocio, ya que corresponden a pedidos que todavía no alcanzaron la etapa de entrega (created, approved, invoiced, processing y shipped) o cuyo proceso fue interrumpido (canceled y unavailable). Sin embargo, se detectaron 8 pedidos con estado delivered que no poseen fecha de entrega al cliente. En 7 de ellos existe una fecha de entrega al transportista, mientras que en 1 tampoco se registra dicha fecha. Estos registros se consideran anomalías de calidad de datos. Debido a que no existe evidencia suficiente para reconstruir la fecha real de entrega, se conservan los valores nulos y no se utiliza order_estimated_delivery_date como sustituto.

In [ ]:
date_columns = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]

for col in date_columns:
    orders[col] = pd.to_datetime(orders[col], errors='coerce')

orders[date_columns].dtypes

order_purchase_timestamp         datetime64[us]
order_approved_at                datetime64[us]
order_delivered_carrier_date     datetime64[us]
order_delivered_customer_date    datetime64[us]
order_estimated_delivery_date    datetime64[us]
dtype: object

Conversión de columnas temporales a tipo datetime
Las columnas relacionadas con fechas y horas se encontraban originalmente almacenadas como tipo str. Para poder realizar operaciones temporales, comparaciones cronológicas y cálculos de duración, se convirtieron al tipo datetime utilizando pd.to_datetime().

Se transformaron las siguientes columnas: order_purchase_timestamp, order_approved_at, order_delivered_carrier_date, order_delivered_customer_date y order_estimated_delivery_date.

Se utilizó el parámetro errors="coerce" para que cualquier valor no interpretable como fecha sea convertido en NaT, preservando así los valores faltantes dentro de un tipo de dato temporal adecuado. Luego de la conversión, se verificó que las cinco columnas quedaran almacenadas como datetime64[us].

Esta transformación permite continuar con controles de calidad más avanzados, como verificar el orden lógico de los eventos del pedido y calcular tiempos transcurridos entre las distintas etapas del proceso.

In [ ]:
orders["order_status"].value_counts()

# verificacion de la coherencia cronologica de las fechas, por ejemplo, que la fecha de aprobacion no sea posterior a la fecha de entrega al cliente
# purchase <= approved <= carrier <= customer
#regla 1 purchase <= approved


approved_before_purchase = orders[
    (orders["order_approved_at"] < orders["order_purchase_timestamp"])
]
approved_before_purchase.shape[0]
#el resultado es 0, lo que significa que no hay registros donde la fecha de aprobación sea anterior a la fecha de compra, lo cual es coherente.

#regla 2 approved <= carrier
carrier_before_approved = orders[
    (orders["order_delivered_carrier_date"] < orders["order_approved_at"])
]
carrier_before_approved.shape[0]
#el resultado es 1359, lo que significa que hay 1359 registros donde la fecha de entrega por el transportista sea anterior a la fecha de aprobación, lo cual es incoherente.

#regla 3 carrier <= customer
customer_before_carrier = orders[
    (orders["order_delivered_customer_date"] < orders["order_delivered_carrier_date"])
]
customer_before_carrier.shape[0]
#el resultado es 23, lo que significa que hay 23 registros donde la fecha de entrega al cliente sea anterior a la fecha de entrega por el transportista, lo cual es incoherente.


carrier_before_approved = orders[
    orders["order_delivered_carrier_date"] < orders["order_approved_at"]
].copy()

carrier_before_approved["time_difference"] = (
    carrier_before_approved["order_approved_at"]
    - carrier_before_approved["order_delivered_carrier_date"]
)
carrier_before_approved["time_difference"].describe()

count                      1359
mean     1 days 00:45:07.153053
std      4 days 19:18:28.055754
min             0 days 00:00:21
25%      0 days 01:24:55.500000
50%             0 days 17:10:04
75%             1 days 01:57:24
max           171 days 05:15:22
Name: time_difference, dtype: object


Se verificó la coherencia cronológica entre las principales fechas del proceso de los pedidos, considerando como secuencia esperada purchase <= approved <= carrier <= customer. No se encontraron pedidos cuya fecha de aprobación fuera anterior a la fecha de compra. Sin embargo, se detectaron 1359 registros donde la fecha de entrega al transportista es anterior a la aprobación del pedido y 23 registros donde la fecha de entrega al cliente es anterior a la fecha registrada de entrega al transportista. Estos casos se consideran anomalías cronológicas potenciales y requieren un análisis adicional de las diferencias temporales antes de clasificarlos definitivamente como errores de calidad de datos.

Se identificaron 1359 pedidos cuya fecha de entrega al transportista es anterior a la fecha de aprobación. Las diferencias presentan una alta dispersión: van desde 21 segundos hasta aproximadamente 171 días. La mediana es de aproximadamente 17 horas y el 75 % de los casos presenta una diferencia inferior a aproximadamente 1 día y 2 horas. La existencia de valores extremos considerablemente mayores sugiere que las anomalías podrían tener distintas causas, por lo que es necesario analizar la distribución de las diferencias antes de determinar su tratamiento.

In [ ]:
carrier_before_approved["difference_hours"] = (
    carrier_before_approved["time_difference"].dt.total_seconds() / 3600
)

print(
    "Menos de 1 hora:",
    (carrier_before_approved["difference_hours"] < 1).sum()
)

print(
    "Entre 1 y 24 horas:",
    (
        (carrier_before_approved["difference_hours"] >= 1) &
        (carrier_before_approved["difference_hours"] < 24)
    ).sum()
)

print(
    "Entre 1 y 7 días:",
    (
        (carrier_before_approved["difference_hours"] >= 24) &
        (carrier_before_approved["difference_hours"] < 168)
    ).sum()
)

print(
    "7 días o más:",
    (carrier_before_approved["difference_hours"] >= 168).sum()
)

carrier_before_approved[
    [
        "order_id",
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "time_difference"
    ]
].sort_values(
    "time_difference",
    ascending=False
).head(10)

customer_before_carrier = orders[
    orders["order_delivered_customer_date"] <
    orders["order_delivered_carrier_date"]
].copy()

customer_before_carrier["time_difference"] = (
    customer_before_carrier["order_delivered_carrier_date"]
    - customer_before_carrier["order_delivered_customer_date"]
)

customer_before_carrier["time_difference"].describe()

Menos de 1 hora: 242
Entre 1 y 24 horas: 659
Entre 1 y 7 días: 444
7 días o más: 14


,order_id,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,time_difference
25883,7c48bb55e8e4f7e56d412e9653db37bc,2018-07-16 18:40:53,2018-07-16 18:50:22,2018-01-26 13:35:00,171 days 05:15:22
14562,1fab4ac9d85079b3da72a11475ae1685,2017-09-01 19:04:22,2017-09-13 22:06:11,2017-09-04 13:10:23,9 days 08:55:48
46163,0184d4ddb259e1a4cfc2871888cf97b8,2017-09-01 20:04:28,2017-09-13 22:17:15,2017-09-04 14:05:50,9 days 08:11:25
98710,1378f9601350615613cc8832d6789c5d,2017-09-01 20:28:02,2017-09-13 22:03:51,2017-09-04 18:07:55,9 days 03:55:56
41592,8554cb37f7158cb0b082a841d24a4589,2017-09-01 18:40:44,2017-09-13 21:58:04,2017-09-04 19:12:19,9 days 02:45:45
11738,cf72398d0690f841271b695bbfda82d2,2017-09-01 18:45:33,2017-09-13 22:04:39,2017-09-04 20:12:41,9 days 01:51:58
85393,40de47dfa620d667117e4a6067b6e1ec,2017-09-01 20:05:55,2017-09-13 21:58:38,2017-09-04 20:36:58,9 days 01:21:40
31211,bc4854efd86d9f42140c951c595d20c1,2017-09-01 20:05:42,2017-09-13 22:00:51,2017-09-04 20:49:57,9 days 01:10:54
55302,77ca435b03fbf991e5027e3776e37885,2017-09-01 18:49:54,2017-09-13 21:56:08,2017-09-04 20:50:00,9 days 01:06:08
68315,580f3268bc2075c5961b2d2929c7a35b,2017-09-01 20:17:09,2017-09-13 22:03:42,2017-09-05 11:58:18,8 days 10:05:24


Análisis de coherencia cronológica

Se verificó la secuencia temporal esperada del proceso:

purchase <= approved <= carrier <= customer

No se detectaron pedidos cuya fecha de aprobación fuera anterior a la fecha de compra.

Se encontraron 1359 registros donde order_delivered_carrier_date es anterior a order_approved_at. Al analizar la magnitud de las diferencias:

242 casos (17,8%) presentan diferencias menores a 1 hora.
659 casos (48,5%) presentan diferencias entre 1 y 24 horas.
444 casos (32,7%) presentan diferencias entre 1 y 7 días.
14 casos (1,0%) presentan diferencias iguales o superiores a 7 días.

Aproximadamente el 66,3% de estas anomalías presenta una diferencia inferior a 24 horas, por lo que podrían existir factores relacionados con el registro o sincronización de eventos. Sin embargo, los casos con diferencias de varios días, y especialmente los valores extremos, requieren ser considerados anomalías de calidad.

También se detectaron 23 pedidos donde la fecha de entrega al cliente es anterior a la fecha de entrega al transportista. Las diferencias varían entre aproximadamente 23 minutos y 16 días, con una mediana cercana a 1 día y 16 horas. Dado que estos registros contradicen directamente el orden lógico del proceso y no existe información suficiente para reconstruir las fechas correctas, se conservan sin modificación y se documentan como anomalías cronológicas.

In [ ]:
#filas completamente duplicadas
orders.duplicated().sum()

#el resultado es 0, lo que significa que no hay filas completamente duplicadas en el DataFrame de órdenes, lo cual es un buen indicador de la calidad de los datos.

#duplicados de la posible clave primaria, que es order_id
orders["order_id"].duplicated().sum()
#el resultado es 0, lo que significa que no hay duplicados en la columna order_id, lo cual es un buen indicador de la calidad de los datos y sugiere que order_id puede ser una clave primaria adecuada para el DataFrame de órdenes.

#verificacion de los valores nulos en la posible clave primaria, que es order_id
orders["order_id"].isna().sum()
#el resultado es 0, lo que significa que no hay valores nulos en la columna order_id, lo cual es un buen indicador de la calidad de los datos y sugiere que order_id puede ser una clave primaria adecuada para el DataFrame de órdenes.

np.int64(0)

Verificación de duplicados y clave primaria

Se verificó la existencia de registros completamente duplicados mediante duplicated(), sin encontrarse filas repetidas.

Posteriormente se analizó order_id, identificador de cada pedido. No se detectaron valores duplicados en esta columna. Además, se verificó la ausencia de valores nulos.

Por lo tanto, order_id presenta valores únicos y completos para todos los registros, por lo que cumple las condiciones necesarias para funcionar como clave primaria del dataset orders.

In [ ]:
print("Nulos en order_id:", orders["order_id"].isna().sum())

print("Nulos en customer_id:", orders["customer_id"].isna().sum())

print("Customer_id duplicados:", orders["customer_id"].duplicated().sum())

customers = tables["olist_customers_dataset"].copy()
customers.info()

customers_not_found = orders[~orders["customer_id"].isin(customers["customer_id"])]
print("Clientes no encontrados:", customers_not_found.shape[0])

customers_without_orders = customers[~customers["customer_id"].isin(orders["customer_id"])]
print("Clientes sin órdenes:", customers_without_orders.shape[0])

print(
    "Nulos customer_id en customers:", customers["customer_id"].isna().sum()
)

print(
    "Duplicados customer_id en customers:", customers["customer_id"].duplicated().sum()
)

print(
    "customer_id unicos en customers:", customers["customer_id"].nunique()
)

print(
    "customer_unique_id en orders:", orders["customer_id"].nunique()
)

print(
    "customer_unique_id únicos:",
    customers["customer_unique_id"].nunique()
)

print(
    "customer_unique_id duplicados:",
    customers["customer_unique_id"].duplicated().sum()
)

customers["customer_unique_id"].value_counts().head(10)


Nulos en order_id: 0
Nulos en customer_id: 0
Customer_id duplicados: 0
<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   customer_id               99441 non-null  str  
 1   customer_unique_id        99441 non-null  str  
 2   customer_zip_code_prefix  99441 non-null  int64
 3   customer_city             99441 non-null  str  
 4   customer_state            99441 non-null  str  
dtypes: int64(1), str(4)
memory usage: 11.0 MB
Clientes no encontrados: 0
Clientes sin órdenes: 0
Nulos customer_id en customers: 0
Duplicados customer_id en customers: 0
customer_id unicos en customers: 99441
customer_unique_id en orders: 99441
customer_unique_id únicos: 96096
customer_unique_id duplicados: 3345


customer_unique_id
8d50f5eadf50201ccdcedfb9e2ac8455    17
3e43e6105506432c953e165fb2acf44c     9
1b6c7548a2a1f9037c1fd3ddfed95f33     7
6469f99c1f9dfae7733b25662e7f1782     7
ca77025e7201e3b30c44b472ff346268     7
47c1a3033b8b77b3ab6e109eb4d5fdf3     6
dc813062e0fc23409cd255f7f53c7074     6
63cfc61cee11cbe306bff5857d00bfe4     6
f0e310a6839dce9de1638e0fe5ab282a     6
de34b16117594161a6a89c50b289d35a     6
Name: count, dtype: int64

Verificación de claves e integridad referencial

Se verificó que order_id no contiene valores nulos ni duplicados, por lo que cumple las condiciones necesarias para identificar de forma única cada pedido en orders.

Asimismo, customer_id no presenta valores nulos ni duplicados tanto en orders como en customers.

Para comprobar la integridad referencial entre ambos datasets, se verificó que todos los customer_id presentes en orders existieran en customers. No se encontraron referencias inexistentes. También se comprobó que todos los registros de customers poseen un pedido asociado.

Por lo tanto, la relación mediante customer_id presenta integridad referencial completa en los datos analizados.

Aunque customer_id presenta un valor único para cada registro del dataset customers, se identificaron 96.096 valores únicos de customer_unique_id sobre un total de 99.441 registros.

Se observaron 3.345 apariciones adicionales de customer_unique_id, lo cual no representa un problema de duplicación. customer_id identifica al registro de cliente asociado a cada pedido, mientras que customer_unique_id permite reconocer a un mismo cliente a través de diferentes pedidos.

Por lo tanto, la repetición de customer_unique_id constituye información válida del negocio y permitirá posteriormente analizar clientes recurrentes y frecuencia de compra.

In [ ]:
customer_order_counts = customers["customer_unique_id"].value_counts()

print(
    "Clientes con más de una compra:",
    (customer_order_counts > 1).sum()
)

print(
    "Clientes con una sola compra:",
    (customer_order_counts == 1).sum()
)

print(
    "Máximo de compras de un cliente:",
    customer_order_counts.max()
)

Clientes con más de una compra: 2997
Clientes con una sola compra: 93099
Máximo de compras de un cliente: 17


De los 96.096 clientes únicos identificados, 2.997 realizaron más de una compra, mientras que 93.099 realizaron una sola. El máximo observado fue de 17 pedidos asociados a un mismo cliente.


In [ ]:
orders_items = tables["olist_order_items_dataset"].copy()
orders_items.info()
orders_items.head()

print(
    "numero de order_id duplicados en orders_items:",
    orders_items["order_id"].duplicated().sum()
)

print(
    "numero de order_id nulos en orders_items:",
    orders_items["order_id"].isna().sum()
)

print(
    "numero de order_id unicos en orders_items:",
    orders_items["order_id"].nunique()
)

print(
    "numero de order_item_id duplicados en orders_items:",
    orders_items["order_item_id"].duplicated().sum()
)

print(
    "numero de order_item_id nulos en orders_items:",
    orders_items["order_item_id"].isna().sum()
)

print(
    "numero de order_item_id unicos en orders_items:",
    orders_items["order_item_id"].nunique()
)

print(
    "numero de product_id duplicados en orders_items:",
    orders_items["product_id"].duplicated().sum()
)

print(
    "numero de product_id nulos en orders_items:",
    orders_items["product_id"].isna().sum()
)

print(
    "numero de product_id unicos en orders_items:",
    orders_items["product_id"].nunique()
)

order_id_not_in_orders = orders_items[~orders_items["order_id"].isin(orders["order_id"])]
print(
    "Número de order_id en orders_items que no están en orders:",
    order_id_not_in_orders.shape[0]
)

#comprobacion de la posible clave primaria compuesta en orders_items, que sería (order_id, order_item_id)

print(
    "Número de duplicados en la combinación (order_id, order_item_id):",
    orders_items.duplicated(subset=["order_id", "order_item_id"]).sum()
)

print(
    "Número de nulos en la combinación (order_id, order_item_id):",
    orders_items[["order_id", "order_item_id"]].isna().sum().sum()
)

orders_whitout_items = orders[~orders["order_id"].isin(orders_items["order_id"])]
print(
    "Número de órdenes en orders que no tienen items en orders_items:",
    orders_whitout_items.shape[0]
)
orders_whitout_items.value_counts("order_status")

<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  str    
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  str    
 3   seller_id            112650 non-null  str    
 4   shipping_limit_date  112650 non-null  str    
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), str(4)
memory usage: 18.4 MB
numero de order_id duplicados en orders_items: 13984
numero de order_id nulos en orders_items: 0
numero de order_id unicos en orders_items: 98666
numero de order_item_id duplicados en orders_items: 112629
numero de order_item_id nulos en orders_items: 0
numero de order_item_id unicos en orders_items: 21
numero de product_id duplicados en orders_items: 79699
numero de product_id nu

order_status
unavailable    603
canceled       164
created          5
invoiced         2
shipped          1
Name: count, dtype: int64

Verificación de claves e integridad entre orders y order_items

El dataset order_items contiene múltiples registros asociados a un mismo order_id, lo cual es coherente con el modelo de negocio, ya que una orden puede contener más de un ítem. Por este motivo, order_id no identifica individualmente cada fila de esta tabla.

Se evaluó la combinación (order_id, order_item_id) como posible clave compuesta. No se encontraron valores nulos ni combinaciones duplicadas, por lo que esta combinación permite identificar de forma única cada ítem dentro de una orden.

También se verificó la integridad referencial desde order_items hacia orders, comprobando que todos los order_id presentes en order_items existen en el dataset orders. No se encontraron referencias a pedidos inexistentes.

Al comprobar la relación en sentido inverso se detectaron 775 pedidos de orders que no poseen registros asociados en order_items. Su distribución por estado es:

unavailable: 603
canceled: 164
created: 5
invoiced: 2
shipped: 1

La mayor parte de estos casos es consistente con pedidos que no completaron el flujo normal de procesamiento, principalmente aquellos con estado unavailable o canceled. Los registros created e invoiced también corresponden a etapas tempranas del ciclo del pedido.

Sin embargo, se identificó un pedido con estado shipped que no posee ningún ítem asociado, lo cual contradice el comportamiento esperado del proceso y se conserva como una anomalía potencial para su posterior revisión. No se modifica el registro debido a que no existe evidencia suficiente para reconstruir la información faltante

Clave de order_items:
(order_id, order_item_id)

Duplicados de la clave compuesta: 0
Nulos en la clave compuesta: 0

Integridad order_items → orders:
0 referencias inexistentes

Orders sin items:
775

In [ ]:
#integridad referencial entre orders_items, products y sellers
products = tables["olist_products_dataset"].copy()
sellers = tables["olist_sellers_dataset"].copy()

product_not_found = orders_items[
    ~orders_items["product_id"].isin(products["product_id"])
]

seller_not_found = orders_items[
    ~orders_items["seller_id"].isin(sellers["seller_id"])
]

print(
    "product_id de order_items que no existen en products:",
    product_not_found.shape[0]
)

print(
    "seller_id de order_items que no existen en sellers:",
    seller_not_found.shape[0]
)

print(
    "filas completamente duplicadas en orders_items:",
    orders_items.duplicated().sum()
)
#conversion de str a datetime de la columna shipping_limit_date en orders_items
orders_items["shipping_limit_date"] = pd.to_datetime(orders_items["shipping_limit_date"], errors='coerce')
orders_items["shipping_limit_date"].dtypes
orders_items["shipping_limit_date"].isna().sum()



product_id de order_items que no existen en products: 0
seller_id de order_items que no existen en sellers: 0
filas completamente duplicadas en orders_items: 0


np.int64(0)

### Integridad referencial y preparación de `order_items`

Se verificó la integridad referencial de las claves foráneas `product_id` y `seller_id`.

Todos los `product_id` presentes en `order_items` existen en el dataset `products`, y todos los `seller_id` se encuentran registrados en `sellers`. Por lo tanto, no se detectaron referencias a productos o vendedores inexistentes.

También se comprobó la existencia de filas completamente duplicadas en `order_items`, obteniéndose 0 registros duplicados.

Finalmente, la columna `shipping_limit_date`, originalmente almacenada como texto, fue convertida a tipo `datetime` mediante `pd.to_datetime()`. Se utilizó `errors="coerce"` para transformar posibles valores no interpretables en `NaT`.

Luego de la conversión no se generaron valores nulos adicionales, lo que indica que todas las fechas pudieron interpretarse correctamente.

In [ ]:
#evaluacion de calidad de datos de price
orders_items["price"].describe()
print(
    "Precios negativos:",
    (orders_items["price"] < 0).sum()
)

print(
    "Precios iguales a 0:",
    (orders_items["price"] == 0).sum()
)

orders_items[
    [
        "order_id",
        "product_id",
        "price"
    ]
].sort_values(
    "price",
    ascending=False
).head(10)

Precios negativos: 0
Precios iguales a 0: 0


,order_id,product_id,price
3556,0812eb902a67711a1cb742b3cdaa65ae,489ae2aa008f021502940f251d4cce7f,6735.00
112233,fefacc66af859508bf1a7934eab1e97f,69c590f7ffc7bf8db97190b6cb6ed62e,6729.00
107841,f5136e38d1a14a4dbd87dff67da82701,1bdf5e6731585cf01aa8169c7028d6ad,6499.00
74336,a96610ab360d42a2e5335a3998b4718a,a6492cc69376c469ab6f61d8f44de961,4799.00
11249,199af31afc78c699f0dbf71fb178d4d4,c3ed642d592594bb648ff4a04cee2747,4690.00
62086,8dbc85d1447242f3b127dda390d56e19,259037a6a41845e455183f89c5035f18,4590.00
29193,426a9742b533fc6fed17d1fd6d143d7e,a1beef8f3992dbd4cd8726796aa69c53,4399.87
45843,68101694e5c5dc7330c91e1bbc36214f,6cdf8fc1d741c76586d8b6b15e9eef30,4099.99
78310,b239ca7cd485940b31882363b52e6674,dd113cb02b2af9c8e5787e8f1f0722f6,4059.00
59137,86c4eab1571921a6a6e248ed312f5a5a,6902c1962dd19d540807d0ab8fade5c6,3999.90


### Validación de `price`

Se analizó la variable `price` para identificar valores económicamente imposibles o potencialmente anómalos.

No se encontraron precios negativos ni productos con precio igual a cero, por lo que todos los registros presentan valores positivos.

También se inspeccionaron los precios más elevados del dataset. Si bien existen productos con valores considerablemente superiores al resto, un precio elevado no constituye por sí mismo un error de calidad. Estos registros se conservan, ya que no existe evidencia suficiente para considerarlos incorrectos.

Los valores extremos podrán ser analizados posteriormente durante el análisis exploratorio de datos (EDA), donde se estudiará la distribución de precios y su impacto sobre las métricas del negocio.

In [ ]:
#calidad de freights, que es la columna freight_value en orders_items
orders_items["freight_value"].describe()

print(
    "Fretes negativos:",
    (orders_items["freight_value"] < 0).sum()
)

print(
    "Fretes iguales a 0:",
    (orders_items["freight_value"] == 0).sum()
)

Fretes negativos: 0
Fretes iguales a 0: 383


### Validación de `freight_value`

Se analizó la variable `freight_value`, correspondiente al valor del flete asociado a cada ítem.

No se detectaron valores negativos, por lo que no existen importes de envío económicamente imposibles.

Se encontraron 383 registros con `freight_value` igual a cero. Estos casos no se consideran automáticamente errores de calidad, ya que pueden corresponder a envíos gratuitos, promociones u otras condiciones comerciales.

Debido a que no existe evidencia suficiente para considerar estos valores incorrectos, se conservan sin modificación. Su comportamiento podrá analizarse posteriormente durante el análisis exploratorio.

In [ ]:
#coherencia minima de shipping_limit_date, que es la fecha limite de envio, con order_purchase_timestamp, que es la fecha de compra
orders_items_check = orders_items.merge(
    orders[
        ["order_id", "order_purchase_timestamp"]
    ],
    on="order_id",
    how="left"
)

shipping_before_purchase = orders_items_check[
    orders_items_check["shipping_limit_date"]
    <
    orders_items_check["order_purchase_timestamp"]
]

shipping_before_purchase.shape[0]

0

### Coherencia temporal de `shipping_limit_date`

Se verificó la coherencia cronológica entre `shipping_limit_date` y `order_purchase_timestamp`.

Dado que `shipping_limit_date` representa la fecha límite establecida para el despacho del pedido, se espera que esta fecha sea igual o posterior al momento en que se realizó la compra.

Para realizar la comprobación se relacionó `order_items` con `orders` mediante `order_id` y se buscaron registros donde:

`shipping_limit_date < order_purchase_timestamp`

No se encontraron casos que incumplieran esta condición. Por lo tanto, las fechas límite de envío presentan coherencia respecto de las fechas de compra registradas.

## Conclusión de la auditoría de `order_items`

El dataset `order_items` presenta una estructura consistente y adecuada para continuar con el análisis.

La combinación `(order_id, order_item_id)` permite identificar de forma única cada ítem de una orden, sin valores nulos ni duplicados.

Se verificó correctamente la integridad referencial de:

- `order_id` con `orders`.
- `product_id` con `products`.
- `seller_id` con `sellers`.

Se detectaron 775 pedidos de `orders` sin registros asociados en `order_items`. La mayoría corresponde a estados coherentes con pedidos que no completaron el proceso (`unavailable`, `canceled`, `created` o `invoiced`). Se identificó un único pedido con estado `shipped` sin ítems asociados, que se conserva documentado como anomalía potencial.

La columna `shipping_limit_date` fue convertida correctamente a `datetime` y no se encontraron fechas límite anteriores a la compra.

Las variables económicas `price` y `freight_value` no presentan valores negativos. `price` tampoco contiene valores iguales a cero, mientras que se encontraron 383 registros con flete igual a cero, considerados potencialmente válidos debido a posibles condiciones de envío gratuito.

No se realizaron imputaciones ni eliminaciones adicionales, ya que no existe evidencia que justifique modificar estos registros.

In [ ]:
payments = tables["olist_order_payments_dataset"].copy()

payments.info()

print("\nFilas duplicadas:", payments.duplicated().sum())

print("\nNulos por columna:")
print(payments.isna().sum())

print("\nOrder_id únicos:", payments["order_id"].nunique())
print("Order_id duplicados:", payments["order_id"].duplicated().sum())

print(
    "\nDuplicados (order_id, payment_sequential):",
    payments.duplicated(
        subset=["order_id", "payment_sequential"]
    ).sum()
)

print(
    "Nulos (order_id, payment_sequential):",
    payments[
        ["order_id", "payment_sequential"]
    ].isna().sum().sum()
)


payment_order_not_found = payments[
    ~payments["order_id"].isin(orders["order_id"])
]

orders_without_payments = orders[
    ~orders["order_id"].isin(payments["order_id"])
]

print(
    "Pagos cuyo order_id no existe en orders:",
    payment_order_not_found.shape[0]
)

print(
    "Orders sin pagos:",
    orders_without_payments.shape[0]
)

print("\nEstados de orders sin pagos:")
print(orders_without_payments["order_status"].value_counts())


print("\nTipos de pago:")
print(payments["payment_type"].value_counts())

print("\nPayment value:")
print(payments["payment_value"].describe())

print(
    "Pagos negativos:",
    (payments["payment_value"] < 0).sum()
)

print(
    "Pagos iguales a 0:",
    (payments["payment_value"] == 0).sum()
)

print("\nCuotas:")
print(payments["payment_installments"].describe())

print(
    "Cuotas negativas:",
    (payments["payment_installments"] < 0).sum()
)

print(
    "Cuotas iguales a 0:",
    (payments["payment_installments"] == 0).sum()
)

payments[
    payments["payment_type"] == "not_defined"
]

not_defined_payments = payments[
    payments["payment_type"] == "not_defined"
]

not_defined_payments

payments[
    payments["order_id"].isin(
        not_defined_payments["order_id"]
    )
].sort_values(
    ["order_id", "payment_sequential"]
)

payments[
    payments["payment_value"] == 0
][
    [
        "order_id",
        "payment_sequential",
        "payment_type",
        "payment_installments",
        "payment_value"
    ]
]

zero_payment_orders = payments[
    payments["payment_value"] == 0
]["order_id"]

payments[
    payments["order_id"].isin(zero_payment_orders)
].sort_values(
    ["order_id", "payment_sequential"]
)

payments[
    payments["payment_installments"] == 0
]

<class 'pandas.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  str    
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  str    
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), str(2)
memory usage: 8.1 MB

Filas duplicadas: 0

Nulos por columna:
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

Order_id únicos: 99440
Order_id duplicados: 4446

Duplicados (order_id, payment_sequential): 0
Nulos (order_id, payment_sequential): 0
Pagos cuyo order_id no existe en orders: 0
Orders sin pagos: 1

Estados de orders sin pagos:
order_status
delivered    1
Name: count, dtype: int64

Tipos de pa

,order_id,payment_sequential,payment_type,payment_installments,payment_value
46982,744bade1fcf9ff3f31d860ace076d422,2,credit_card,0,58.69
79014,1a57108394169c0b47d8f876acc9ba2d,2,credit_card,0,129.94


Se detectaron 9 registros con payment_value = 0. La mayoría corresponde al método de pago voucher, mientras que 3 presentan payment_type = "not_defined". La asociación entre pagos de valor cero y vouchers sugiere que estos registros podrían responder a una lógica similar; sin embargo, no existe evidencia suficiente para reclasificar los casos not_defined. Por este motivo, se conservan los valores originales y se documentan como casos atípicos sin impacto monetario registrado.

### Análisis de pagos con valor cero

Se identificaron 9 registros con `payment_value = 0`, de los cuales varios corresponden al método de pago `voucher` y 3 presentan `payment_type = "not_defined"`.

Para determinar si estos registros representaban pedidos sin pago, se analizaron todos los pagos asociados a los mismos `order_id`. Se comprobó que dichos pedidos poseen otros registros con valores de pago positivos.

Por lo tanto, los registros con valor cero no representan necesariamente órdenes impagas, sino que forman parte de pedidos con múltiples registros de pago. En algunos casos se combinan vouchers con otros medios de pago.

Los 3 registros `not_defined` también presentan valor cero y están asociados a pedidos con pagos positivos. Aunque su comportamiento es similar al observado en algunos vouchers, no existe evidencia suficiente para reclasificarlos, por lo que se conserva su categoría original.

No se realizan modificaciones sobre estos registros.

### Validación de `payment_installments`

Se verificaron los valores de `payment_installments` para detectar cantidades de cuotas imposibles o inconsistentes.

No se encontraron valores negativos. Sin embargo, se identificaron 2 registros con `payment_installments = 0`.

Ambos corresponden a pagos mediante `credit_card` y presentan valores monetarios positivos (`payment_value > 0`). Dado que un pago efectivo con tarjeta de crédito debería registrar al menos una cuota, estos casos se consideran anomalías de calidad de datos.

No se modifica el número de cuotas debido a que no existe evidencia suficiente para determinar el valor correcto.

In [ ]:
products = tables["olist_products_dataset"].copy()

products.info()

print("\nFilas duplicadas:", products.duplicated().sum())

print("\nNulos por columna:")
print(products.isna().sum())

print(
    "\nproduct_id duplicados:",
    products["product_id"].duplicated().sum()
)

print(
    "product_id nulos:",
    products["product_id"].isna().sum()
)

print(
    "product_id únicos:",
    products["product_id"].nunique()
)


products_without_items = products[
    ~products["product_id"].isin(orders_items["product_id"])
]

print(
    "Productos sin registros en order_items:",
    products_without_items.shape[0]
)


numeric_product_columns = [
    "product_name_lenght",
    "product_description_lenght",
    "product_photos_qty",
    "product_weight_g",
    "product_length_cm",
    "product_height_cm",
    "product_width_cm"
]

for col in numeric_product_columns:
    print(
        f"{col} <= 0:",
        (products[col] <= 0).sum()
    )

product_metadata_nulls = products[
    products["product_category_name"].isna()
]

product_metadata_nulls[
    [
        "product_category_name",
        "product_name_lenght",
        "product_description_lenght",
        "product_photos_qty"
    ]
].isna().sum()

products[
    [
        "product_category_name",
        "product_name_lenght",
        "product_description_lenght",
        "product_photos_qty"
    ]
].isna().sum(axis=1).value_counts()


physical_nulls = products[
    products[
        [
            "product_weight_g",
            "product_length_cm",
            "product_height_cm",
            "product_width_cm"
        ]
    ].isna().any(axis=1)
]

physical_nulls[
    [
        "product_id",
        "product_category_name",
        "product_weight_g",
        "product_length_cm",
        "product_height_cm",
        "product_width_cm"
    ]
]

invalid_weight = products[
    products["product_weight_g"] <= 0
]

invalid_weight[
    [
        "product_id",
        "product_category_name",
        "product_weight_g",
        "product_length_cm",
        "product_height_cm",
        "product_width_cm"
    ]
]

print("Peso <= 0:", (products["product_weight_g"] <= 0).sum())
print("Largo <= 0:", (products["product_length_cm"] <= 0).sum())
print("Alto <= 0:", (products["product_height_cm"] <= 0).sum())
print("Ancho <= 0:", (products["product_width_cm"] <= 0).sum())

<class 'pandas.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  str    
 1   product_category_name       32341 non-null  str    
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), str(2)
memory usage: 3.7 MB

Filas duplicadas: 0

Nulos por columna:
product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g 

## Auditoría del dataset `products`

El dataset `products` contiene 32.951 registros y presenta una estructura consistente en cuanto a su identificador principal.

La columna `product_id` no contiene valores nulos ni duplicados, por lo que cumple las condiciones necesarias para identificar de forma única cada producto. También se verificó que todos los productos presentes en este dataset aparecen referenciados desde `order_items`, por lo que no se detectaron productos aislados respecto del flujo principal de ventas.

No se encontraron filas completamente duplicadas.

### Valores faltantes

Se detectaron 610 valores nulos en las siguientes columnas:

- `product_category_name`
- `product_name_lenght`
- `product_description_lenght`
- `product_photos_qty`

Se comprobó que estos cuatro campos faltan simultáneamente en exactamente las mismas 610 filas. Por lo tanto, no se interpretan como cuatro problemas independientes, sino como un mismo patrón de productos sin metadatos descriptivos.

También se detectaron 2 productos con valores faltantes en las variables físicas:

- `product_weight_g`
- `product_length_cm`
- `product_height_cm`
- `product_width_cm`

Estos registros se conservan sin imputación, ya que no existe información suficiente para reconstruir sus características físicas.

### Validación de valores físicos

Se revisaron valores no positivos en las variables físicas del producto.

No se encontraron valores menores o iguales a cero en:

- `product_length_cm`
- `product_height_cm`
- `product_width_cm`

Sin embargo, se identificaron 4 productos con `product_weight_g <= 0`. Dado que un producto físico debería registrar un peso positivo, estos casos se consideran anomalías potenciales de calidad de datos.

No se modifican estos valores, ya que no existe evidencia suficiente para determinar el peso correcto.

### Conclusión

El dataset `products` presenta una estructura general consistente y una clave primaria válida. Los principales hallazgos corresponden a 610 productos sin metadatos descriptivos, 2 productos con información física incompleta y 4 productos con peso no positivo.

Estos registros se conservan y se documentan como casos de calidad de datos, evitando imputaciones sin evidencia.

In [ ]:
sellers = tables["olist_sellers_dataset"].copy()

sellers.info()

print("\nFilas duplicadas:", sellers.duplicated().sum())

print("\nNulos por columna:")
print(sellers.isna().sum())

print(
    "\nseller_id duplicados:",
    sellers["seller_id"].duplicated().sum()
)

print(
    "seller_id nulos:",
    sellers["seller_id"].isna().sum()
)

print(
    "seller_id únicos:",
    sellers["seller_id"].nunique()
)

sellers_without_items = sellers[
    ~sellers["seller_id"].isin(orders_items["seller_id"])
]

print(
    "Sellers sin registros en order_items:",
    sellers_without_items.shape[0]
)

print(
    "ZIP nulos:",
    sellers["seller_zip_code_prefix"].isna().sum()
)

print(
    "Ciudad nula:",
    sellers["seller_city"].isna().sum()
)

print(
    "Estado nulo:",
    sellers["seller_state"].isna().sum()
)

print("\nEstados:")
print(sellers["seller_state"].value_counts())

<class 'pandas.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype
---  ------                  --------------  -----
 0   seller_id               3095 non-null   str  
 1   seller_zip_code_prefix  3095 non-null   int64
 2   seller_city             3095 non-null   str  
 3   seller_state            3095 non-null   str  
dtypes: int64(1), str(3)
memory usage: 230.3 KB

Filas duplicadas: 0

Nulos por columna:
seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

seller_id duplicados: 0
seller_id nulos: 0
seller_id únicos: 3095
Sellers sin registros en order_items: 0
ZIP nulos: 0
Ciudad nula: 0
Estado nulo: 0

Estados:
seller_state
SP    1849
PR     349
MG     244
SC     190
RJ     171
RS     129
GO      40
DF      30
ES      23
BA      19
CE      13
PE       9
PB       6
RN       5
MS       5
MT       4
RO       2
SE       2
AC       1
PI 

## Auditoría del dataset `sellers`

El dataset `sellers` contiene 3.095 registros y presenta una estructura consistente.

La columna `seller_id` no contiene valores nulos ni duplicados, por lo que puede utilizarse como clave primaria del dataset.

No se detectaron filas completamente duplicadas ni valores faltantes en las columnas:

- `seller_id`
- `seller_zip_code_prefix`
- `seller_city`
- `seller_state`

También se verificó la relación entre `sellers` y `order_items`. Todos los vendedores registrados en `sellers` poseen al menos un registro asociado en `order_items`, y previamente se comprobó que todos los `seller_id` presentes en `order_items` existen en `sellers`.

Por lo tanto, la relación:

`sellers.seller_id -> order_items.seller_id`

presenta integridad referencial completa en los datos analizados.

La distribución por estado muestra una fuerte concentración de vendedores en determinados estados, especialmente São Paulo (`SP`). Esta distribución no representa un problema de calidad y será considerada posteriormente como información de negocio durante el análisis exploratorio.

### Conclusión

No se detectaron problemas estructurales, valores faltantes, duplicados ni inconsistencias referenciales relevantes en `sellers`.

Por lo tanto, el dataset se considera validado para continuar con las siguientes etapas del análisis.

In [29]:
reviews = tables["olist_order_reviews_dataset"].copy()
reviews.info()

print("\nFilas duplicadas:", reviews.duplicated().sum())
print("\nNulos por columna:")
print(reviews.isna().sum())
#hay nulos por que un cliente puede dejar una review sin dejar un comentario, o puede no dejar una reseña en absoluto. Por lo tanto, es normal que haya nulos en las columnas de review_comment_message y review_score.

print("review_id nulos:", reviews["review_id"].isna().sum())
print("review_id duplicados:", reviews["review_id"].duplicated().sum())
print("review_id únicos:", reviews["review_id"].nunique())

print("\norder_id nulos:", reviews["order_id"].isna().sum())
print("order_id duplicados:", reviews["order_id"].duplicated().sum())
print("order_id únicos:", reviews["order_id"].nunique())

review_order_not_found = reviews[
    ~reviews["order_id"].isin(orders["order_id"])
]

orders_without_reviews = orders[
    ~orders["order_id"].isin(reviews["order_id"])
]

print(
    "Reviews cuyo order_id no existe en orders:",
    review_order_not_found.shape[0]
)

print(
    "Orders sin review:",
    orders_without_reviews.shape[0]
)

print("\nEstados de orders sin review:")
print(orders_without_reviews["order_status"].value_counts())


print(reviews["review_score"].value_counts().sort_index())

print(
    "Scores menores a 1:",
    (reviews["review_score"] < 1).sum()
)

print(
    "Scores mayores a 5:",
    (reviews["review_score"] > 5).sum()
)

review_date_columns = [
    "review_creation_date",
    "review_answer_timestamp"
]

for col in review_date_columns:
    reviews[col] = pd.to_datetime(
        reviews[col],
        errors="coerce"
    )

print(reviews[review_date_columns].dtypes)

print("\nNulos después de conversión:")
print(reviews[review_date_columns].isna().sum())

answer_before_creation = reviews[
    reviews["review_answer_timestamp"] <
    reviews["review_creation_date"]
]

print(
    "Respuestas anteriores a la creación de la review:",
    answer_before_creation.shape[0]
)

print(
    "Duplicados (review_id, order_id):",
    reviews.duplicated(
        subset=["review_id", "order_id"]
    ).sum()
)

print(
    "Nulos (review_id, order_id):",
    reviews[
        ["review_id", "order_id"]
    ].isna().sum().sum()
)

<class 'pandas.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype
---  ------                   --------------  -----
 0   review_id                99224 non-null  str  
 1   order_id                 99224 non-null  str  
 2   review_score             99224 non-null  int64
 3   review_comment_title     11568 non-null  str  
 4   review_comment_message   40977 non-null  str  
 5   review_creation_date     99224 non-null  str  
 6   review_answer_timestamp  99224 non-null  str  
dtypes: int64(1), str(6)
memory usage: 17.8 MB

Filas duplicadas: 0

Nulos por columna:
review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64
review_id nulos: 0
review_id duplicados: 814
review_id únicos: 98410

order_id nulos: 0
order_id duplic

El dataset order_reviews no presenta filas completamente duplicadas ni referencias a pedidos inexistentes. Tanto review_id como order_id presentan repeticiones, por lo que se evaluó la combinación (review_id, order_id) como identificador compuesto de los registros.

Los campos review_comment_title y review_comment_message presentan una elevada cantidad de valores nulos. Estos faltantes se consideran coherentes con la naturaleza opcional de los comentarios, ya que una valoración puede registrarse únicamente mediante review_score.

Se identificaron 768 pedidos sin review asociada. Esta ausencia no se considera una anomalía de calidad, debido a que la generación de una reseña depende de la participación voluntaria del cliente.

Todos los valores de review_score se encuentran dentro del rango válido de 1 a 5. Las columnas temporales fueron convertidas correctamente a datetime.

In [30]:
geolocation = tables["olist_geolocation_dataset"].copy()
customers = tables["olist_customers_dataset"].copy()
sellers = tables["olist_sellers_dataset"].copy()
geolocation.info()

print("\nNulos por columna:")
print(geolocation.isna().sum())

print(
    "\nFilas completamente duplicadas:",
    geolocation.duplicated().sum()
)

print(
    "ZIP únicos:",
    geolocation["geolocation_zip_code_prefix"].nunique()
)

print(
    "ZIP duplicados:",
    geolocation["geolocation_zip_code_prefix"].duplicated().sum()
)

print(
    "Latitudes fuera de rango:",
    (
        (geolocation["geolocation_lat"] < -90) |
        (geolocation["geolocation_lat"] > 90)
    ).sum()
)

print(
    "Longitudes fuera de rango:",
    (
        (geolocation["geolocation_lng"] < -180) |
        (geolocation["geolocation_lng"] > 180)
    ).sum()
)

print(
    "Latitud = 0:",
    (geolocation["geolocation_lat"] == 0).sum()
)

print(
    "Longitud = 0:",
    (geolocation["geolocation_lng"] == 0).sum()
)

customer_zip_not_found = customers[
    ~customers["customer_zip_code_prefix"].isin(
        geolocation["geolocation_zip_code_prefix"]
    )
]

seller_zip_not_found = sellers[
    ~sellers["seller_zip_code_prefix"].isin(
        geolocation["geolocation_zip_code_prefix"]
    )
]

print(
    "Customers cuyo ZIP no existe en geolocation:",
    customer_zip_not_found.shape[0]
)

print(
    "Sellers cuyo ZIP no existe en geolocation:",
    seller_zip_not_found.shape[0]
)

print(
    "Estados distintos:",
    geolocation["geolocation_state"].nunique()
)

print(
    geolocation["geolocation_state"]
    .value_counts()
)

geolocation_clean = geolocation.drop_duplicates().copy()

print("Filas originales:", geolocation.shape[0])
print("Filas limpias:", geolocation_clean.shape[0])

<class 'pandas.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  str    
 4   geolocation_state            1000163 non-null  str    
dtypes: float64(2), int64(1), str(2)
memory usage: 50.1 MB

Nulos por columna:
geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

Filas completamente duplicadas: 261831
ZIP únicos: 19015
ZIP duplicados: 981148
Latitudes fuera de rango: 0
Longitudes fuera de rango: 0
Latitud = 0: 0
Longitud = 0: 0
Customers cuyo ZIP no existe en geolocation: 278
Sell

El dataset geolocation no presenta valores nulos ni coordenadas fuera de los rangos geográficos válidos.

Se identificaron múltiples registros asociados a un mismo prefijo postal. Esta repetición es esperable debido a que un código postal puede estar representado por diferentes coordenadas geográficas.

Sin embargo, se detectaron 261.831 filas completamente duplicadas, es decir, registros idénticos en todas las columnas. Estos registros representan redundancia y se eliminan únicamente en una copia destinada al análisis mediante drop_duplicates(), preservando el dataset original.

También se detectaron 278 registros de clientes y 7 registros de vendedores cuyos códigos postales no poseen correspondencia en geolocation. Estos casos se interpretan como una limitación de cobertura del dataset geográfico y no se imputan coordenadas sin evidencia.

In [ ]:
category_translation = tables[
    "product_category_name_translation"
].copy()
products = tables["olist_products_dataset"].copy()

category_translation.info()

print("\nNulos:")
print(category_translation.isna().sum())

print(
    "\nFilas duplicadas:",
    category_translation.duplicated().sum()
)

print(
    "Categorías originales duplicadas:",
    category_translation[
        "product_category_name"
    ].duplicated().sum()
)

print(
    "Categorías traducidas duplicadas:",
    category_translation[
        "product_category_name_english"
    ].duplicated().sum()
)

print(
    "Categorías originales únicas:",
    category_translation[
        "product_category_name"
    ].nunique()
)

product_categories = (
    products["product_category_name"]
    .dropna()
    .unique()
)

categories_without_translation = [
    category
    for category in product_categories
    if category not in set(
        category_translation["product_category_name"]
    )
]

print(
    "Categorías de products sin traducción:",
    len(categories_without_translation)
)

print(categories_without_translation)

translations_without_products = category_translation[
    ~category_translation["product_category_name"].isin(
        products["product_category_name"]
    )
]

print(
    "Traducciones sin productos asociados:",
    translations_without_products.shape[0]
)

<class 'pandas.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   product_category_name          71 non-null     str  
 1   product_category_name_english  71 non-null     str  
dtypes: str(2)
memory usage: 3.5 KB

Nulos:
product_category_name            0
product_category_name_english    0
dtype: int64

Filas duplicadas: 0
Categorías originales duplicadas: 0
Categorías traducidas duplicadas: 0
Categorías originales únicas: 71
Categorías de products sin traducción: 2
['pc_gamer', 'portateis_cozinha_e_preparadores_de_alimentos']
Traducciones sin productos asociados: 0


El dataset product_category_name_translation contiene 71 categorías y no presenta valores nulos, filas duplicadas ni duplicados en las categorías originales o traducidas.

Todas las categorías registradas en la tabla de traducción están asociadas a productos. Sin embargo, se identificaron dos categorías utilizadas en products que no poseen traducción al inglés: pc_gamer y portateis_cozinha_e_preparadores_de_alimentos.

Estas categorías se conservan con su nombre original, ya que no existe una traducción proporcionada por el dataset y no se incorpora información externa durante la fase de limpieza.

In [ ]:
# ============================================================
# CREACIÓN DE DATASETS LIMPIOS PARA EL ANÁLISIS
# ============================================================

# Copias desde los datasets originales
orders_clean = tables["olist_orders_dataset"].copy()
customers_clean = tables["olist_customers_dataset"].copy()
order_items_clean = tables["olist_order_items_dataset"].copy()
payments_clean = tables["olist_order_payments_dataset"].copy()
products_clean = tables["olist_products_dataset"].copy()
sellers_clean = tables["olist_sellers_dataset"].copy()
reviews_clean = tables["olist_order_reviews_dataset"].copy()
geolocation_clean = tables["olist_geolocation_dataset"].copy()
category_translation_clean = tables[
    "product_category_name_translation"
].copy()


# ------------------------------------------------------------
# ORDERS - conversión de fechas
# ------------------------------------------------------------

order_date_columns = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]

for col in order_date_columns:
    orders_clean[col] = pd.to_datetime(
        orders_clean[col],
        errors="coerce"
    )


# ------------------------------------------------------------
# ORDER ITEMS - conversión de shipping_limit_date
# ------------------------------------------------------------

order_items_clean["shipping_limit_date"] = pd.to_datetime(
    order_items_clean["shipping_limit_date"],
    errors="coerce"
)


# ------------------------------------------------------------
# REVIEWS - conversión de fechas
# ------------------------------------------------------------

review_date_columns = [
    "review_creation_date",
    "review_answer_timestamp"
]

for col in review_date_columns:
    reviews_clean[col] = pd.to_datetime(
        reviews_clean[col],
        errors="coerce"
    )


# ------------------------------------------------------------
# GEOLOCATION - eliminación de duplicados exactos
# ------------------------------------------------------------

geolocation_clean = (
    geolocation_clean
    .drop_duplicates()
    .copy()
)


# ------------------------------------------------------------
# PRODUCTS - incorporación de traducciones
# ------------------------------------------------------------

products_clean = products_clean.merge(
    category_translation_clean,
    on="product_category_name",
    how="left"
)

products_clean["category_name"] = (
    products_clean["product_category_name_english"]
    .fillna(products_clean["product_category_name"])
)

print("ORDERS:")
print(orders_clean[order_date_columns].dtypes)

print("\nORDER ITEMS:")
print(order_items_clean["shipping_limit_date"].dtype)

print("\nREVIEWS:")
print(reviews_clean[review_date_columns].dtypes)

print(
    "\nFilas geolocation originales:",
    len(tables["olist_geolocation_dataset"])
)

print(
    "Filas geolocation clean:",
    len(geolocation_clean)
)



ORDERS:
order_purchase_timestamp         datetime64[us]
order_approved_at                datetime64[us]
order_delivered_carrier_date     datetime64[us]
order_delivered_customer_date    datetime64[us]
order_estimated_delivery_date    datetime64[us]
dtype: object

ORDER ITEMS:
datetime64[us]

REVIEWS:
review_creation_date       datetime64[us]
review_answer_timestamp    datetime64[us]
dtype: object

Filas geolocation originales: 1000163
Filas geolocation clean: 738332


In [49]:
orders_analysis = orders_clean.merge(
    customers_clean[
        [
            "customer_id",
            "customer_unique_id",
            "customer_city",
            "customer_state"
        ]
    ],
    on="customer_id",
    how="left"
)

print("Filas:", orders_analysis.shape[0])
print("Order ID únicos:", orders_analysis["order_id"].nunique())

orders_analysis["purchase_year"] = (
    orders_analysis["order_purchase_timestamp"].dt.year
)

orders_analysis["purchase_month"] = (
    orders_analysis["order_purchase_timestamp"].dt.month
)

orders_analysis["purchase_year_month"] = (
    orders_analysis["order_purchase_timestamp"]
    .dt.to_period("M")
    .astype(str)
)

orders_analysis["delivery_time_days"] = (
    orders_analysis["order_delivered_customer_date"]
    - orders_analysis["order_purchase_timestamp"]
).dt.total_seconds() / 86400

orders_analysis["delivery_delay_days"] = (
    orders_analysis["order_delivered_customer_date"]
    - orders_analysis["order_estimated_delivery_date"]
).dt.total_seconds() / 86400

orders_analysis["delivered_late"] = (
    orders_analysis["delivery_delay_days"] > 0
)

orders_analysis["delivered_late"] = pd.NA

mask_delivered_date = (
    orders_analysis["order_delivered_customer_date"].notna()
)

orders_analysis.loc[
    mask_delivered_date,
    "delivered_late"
] = (
    orders_analysis.loc[
        mask_delivered_date,
        "delivery_delay_days"
    ] > 0
)

items_by_order = (
    order_items_clean
    .groupby("order_id")
    .agg(
        items_count=("order_item_id", "count"),
        products_count=("product_id", "nunique"),
        sellers_count=("seller_id", "nunique"),
        products_value=("price", "sum"),
        freight_value=("freight_value", "sum")
    )
    .reset_index()
)

orders_analysis = orders_analysis.merge(
    items_by_order,
    on="order_id",
    how="left"
)

print("Filas:", orders_analysis.shape[0])
print("Order ID únicos:", orders_analysis["order_id"].nunique())

payments_by_order = (
    payments_clean
    .groupby("order_id")
    .agg(
        payment_value=("payment_value", "sum"),
        payment_records=("payment_sequential", "count"),
        max_installments=("payment_installments", "max"),
        payment_types_count=("payment_type", "nunique")
    )
    .reset_index()
)

orders_analysis = orders_analysis.merge(
    payments_by_order,
    on="order_id",
    how="left"
)

print(orders_analysis.shape[0])
print(orders_analysis["order_id"].nunique())


orders_analysis["has_items"] = (
    orders_analysis["items_count"].notna()
)

orders_analysis["has_payment"] = (
    orders_analysis["payment_value"].notna()
)



reviews_by_order = (
    reviews_clean
    .groupby("order_id")
    .agg(
        review_score_mean=("review_score", "mean"),
        review_count=("review_id", "count"),
        review_comments_count=(
            "review_comment_message",
            lambda x: x.notna().sum()
        )
    )
    .reset_index()
)

orders_analysis = orders_analysis.merge(
    reviews_by_order,
    on="order_id",
    how="left"
)

orders_analysis["has_review"] = (
    orders_analysis["review_count"].notna()
)

print("Filas:", orders_analysis.shape[0])
print(
    "Order ID únicos:",
    orders_analysis["order_id"].nunique()
)

orders_analysis.info()

orders_analysis[
    [
        "order_id",
        "customer_unique_id",
        "order_status",
        "items_count",
        "products_value",
        "freight_value",
        "payment_value",
        "delivery_time_days",
        "delivery_delay_days",
        "delivered_late"
    ]
].head(10)

Filas: 99441
Order ID únicos: 99441
Filas: 99441
Order ID únicos: 99441
99441
99441
Filas: 99441
Order ID únicos: 99441
<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 32 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  str           
 1   customer_id                    99441 non-null  str           
 2   order_status                   99441 non-null  str           
 3   order_purchase_timestamp       99441 non-null  datetime64[us]
 4   order_approved_at              99281 non-null  datetime64[us]
 5   order_delivered_carrier_date   97658 non-null  datetime64[us]
 6   order_delivered_customer_date  96476 non-null  datetime64[us]
 7   order_estimated_delivery_date  99441 non-null  datetime64[us]
 8   customer_unique_id             99441 non-null  str           
 9   customer_city                  99441 non

,order_id,customer_unique_id,order_status,items_count,products_value,freight_value,payment_value,delivery_time_days,delivery_delay_days,delivered_late
0,e481f51cbdc54678b7cc49136f2d6af7,7c396fd4830fd04220f754e42b4e5bff,delivered,1.0,29.99,8.72,38.71,8.436574,-7.107488,False
1,53cdb2fc8bc7dce0b6741e2150273451,af07308b275d755c9edb36a90c618231,delivered,1.0,118.70,22.76,141.46,13.782037,-5.355729,False
2,47770eb9100c2d0c44946d9cf07ec65d,3a653a41f6f9fc3d2a113cf8398680e8,delivered,1.0,159.90,19.22,179.12,9.394213,-17.245498,False
3,949d5b44dbf5de918fe9c16f97b45f8a,7c142cf63193a1473d2e66489a9ae977,delivered,1.0,45.00,27.20,72.20,13.208750,-12.980069,False
4,ad21c59c0840e6cb83a9ceb5573f8159,72632f0f9dd73dfee390c9b22eb56dd6,delivered,1.0,19.90,8.72,28.62,2.873877,-9.238171,False
5,a4591c265e18cb1dcee52889e2d8acc3,80bb27c7c16e8f973207a5086ab329e2,delivered,1.0,147.90,27.36,175.26,16.542245,-5.543113,False
6,136cce7faa42fdb2cefd53fdc79a6098,36edbb3fb164b1f16485364b6fb04c73,invoiced,1.0,49.90,16.05,65.95,NaN,NaN,<NA>
7,6514b8ad8028c9f2cc2374ded245783f,932afa1e708222e5821dac9cd5db4cae,delivered,1.0,59.99,15.17,75.16,9.989826,-11.461215,False
8,76c6e866289321a7c93b82b54852dc33,39382392765b6dc74812866ee5ee92a7,delivered,1.0,19.90,16.05,35.95,9.818762,-31.410995,False
9,e69bfb5eb88e0ed6a785585b27e16dbf,299905e3934e9e181bfb2e164dd4b4f8,delivered,1.0,149.99,19.77,169.76,18.221852,-6.281597,False


In [55]:
order_items_analysis = order_items_clean.copy()

print("Filas:", order_items_analysis.shape[0])

order_items_analysis = order_items_analysis.merge(
    products_clean[
        [
            "product_id",
            "category_name",
            "product_weight_g",
            "product_length_cm",
            "product_height_cm",
            "product_width_cm"
        ]
    ],
    on="product_id",
    how="left"
)

order_items_analysis = order_items_analysis.merge(
    sellers_clean[
        [
            "seller_id",
            "seller_city",
            "seller_state"
        ]
    ],
    on="seller_id",
    how="left"
)

order_items_analysis = order_items_analysis.merge(
    orders_analysis[
        [
            "order_id",
            "order_status",
            "order_purchase_timestamp",
            "purchase_year",
            "purchase_month",
            "purchase_year_month",
            "customer_unique_id",
            "customer_city",
            "customer_state"
        ]
    ],
    on="order_id",
    how="left"
)

print("Filas:", order_items_analysis.shape[0])

print(
    "Clave (order_id, order_item_id) única:",
    ~order_items_analysis.duplicated(
        subset=["order_id", "order_item_id"]
    ).any()
)

order_items_analysis["item_total_value"] = (
    order_items_analysis["price"]
    + order_items_analysis["freight_value"]
)

Filas: 112650
Filas: 112650
Clave (order_id, order_item_id) única: True


In [56]:
customer_analysis = (
    orders_analysis
    .groupby("customer_unique_id")
    .agg(
        orders_count=("order_id", "count"),
        total_payment_value=("payment_value", "sum"),
        avg_order_value=("payment_value", "mean"),
        first_purchase=("order_purchase_timestamp", "min"),
        last_purchase=("order_purchase_timestamp", "max"),
        avg_delivery_time_days=("delivery_time_days", "mean"),
        avg_review_score=("review_score_mean", "mean")
    )
    .reset_index()
)

customer_analysis["is_recurring"] = (
    customer_analysis["orders_count"] > 1
)

customer_analysis["customer_lifetime_days"] = (
    customer_analysis["last_purchase"]
    - customer_analysis["first_purchase"]
).dt.total_seconds() / 86400

print("Filas:", customer_analysis.shape[0])

print(
    "Customer únicos:",
    customer_analysis["customer_unique_id"].nunique()
)

print(
    "Clientes recurrentes:",
    customer_analysis["is_recurring"].sum()
)

Filas: 96096
Customer únicos: 96096
Clientes recurrentes: 2997


# Construcción de tablas analíticas

Luego de finalizar la auditoría y preparación de los datasets originales de Olist, se construyeron tablas analíticas con distintas granularidades para facilitar el análisis exploratorio y evitar problemas de duplicación de registros durante los `merge`.

La principal decisión de modelado fue no combinar directamente todas las tablas en un único DataFrame, ya que datasets como `order_items`, `payments` y `reviews` poseen múltiples registros asociados a un mismo pedido. Un `merge` directo entre estas tablas podría multiplicar filas y producir sobreconteos en métricas como ventas, pagos o cantidad de pedidos.

Por este motivo, se construyeron tres tablas analíticas principales.

## `orders_analysis`

Granularidad:

`1 fila = 1 pedido`

La tabla parte de `orders` y se complementa con información agregada de clientes, ítems, pagos y reviews.

Incluye, entre otras variables:

- Identificación del pedido y del cliente.
- Estado del pedido.
- Ciudad y estado del cliente.
- Fechas principales del proceso.
- Año, mes y período de compra.
- Tiempo de entrega.
- Diferencia entre fecha real y fecha estimada de entrega.
- Indicador de entrega tardía.
- Cantidad de ítems.
- Cantidad de productos distintos.
- Cantidad de vendedores asociados.
- Valor total de productos.
- Valor total del flete.
- Valor total pagado.
- Cantidad de registros de pago.
- Número máximo de cuotas.
- Cantidad de medios de pago utilizados.
- Puntuación media de las reviews.
- Cantidad de reviews y comentarios.

Se verificó que la tabla mantuviera la granularidad esperada:

- 99.441 filas.
- 99.441 `order_id` únicos.

Esto confirma que los procesos de agregación y `merge` no generaron duplicación de pedidos.

## `order_items_analysis`

Granularidad:

`1 fila = 1 ítem vendido`

Esta tabla parte de `order_items` y se complementa con información de productos, categorías, vendedores, pedidos y clientes.

Incluye:

- `order_id` y `order_item_id`.
- Producto.
- Categoría del producto.
- Precio.
- Flete.
- Valor total del ítem (`price + freight_value`).
- Peso y dimensiones del producto.
- Vendedor.
- Ciudad y estado del vendedor.
- Estado del pedido.
- Fecha y período de compra.
- Cliente y ubicación del cliente.

La clave compuesta `(order_id, order_item_id)` continúa siendo única luego de los `merge`.

Se verificó:

- 112.650 filas.
- 0 duplicados en `(order_id, order_item_id)`.

Esta tabla será utilizada principalmente para analizar productos, categorías, vendedores, precios, fletes y volumen de ventas.

## `customer_analysis`

Granularidad:

`1 fila = 1 cliente real`

La tabla se construyó agrupando `orders_analysis` por `customer_unique_id`.

Incluye:

- Cantidad total de pedidos.
- Valor total pagado.
- Ticket promedio.
- Primera compra.
- Última compra.
- Tiempo de vida observado del cliente.
- Tiempo promedio de entrega.
- Puntuación media de reviews.
- Indicador de cliente recurrente.

Se verificó:

- 96.096 clientes únicos.
- 96.096 filas.
- 2.997 clientes con más de una compra.

Esta tabla permitirá analizar recurrencia, gasto por cliente y comportamiento general de compra.

## Criterio de modelado

Las tres tablas mantienen granularidades diferentes y complementarias:

- `orders_analysis`: análisis a nivel pedido.
- `order_items_analysis`: análisis a nivel producto vendido.
- `customer_analysis`: análisis a nivel cliente.

Esta separación evita duplicaciones de métricas y permite utilizar la tabla adecuada según la pregunta de negocio que se quiera responder.

Con estas tablas preparadas, la siguiente etapa corresponde al análisis exploratorio de datos (EDA), donde se evaluarán ventas, clientes, productos, logística, reviews y recurrencia.